Improve to_datetime bounds checking

rebecca-palmer opened the pull request:
to_datetime(errors='raise') is supposed to raise an exception on out-of-bounds input.  However, it uses rounding to integer for this bounds checking but not the actual conversion, and hence, for input just outside the bounds it may instead return NaT ([on x86](https://buildd.debian.org/status/fetch.php?pkg=pandas&arch=all&ver=1.5.2%2Bdfsg-1&stamp=1669856039&raw=0)) or output clipped to the bounds ([on arm](https://buildd.debian.org/status/fetch.php?pkg=pandas&arch=arm64&ver=1.5.2%2Bdfsg-1&stamp=1669872650&raw=0)).

It also assumes that converting NaN to int gives NaT, which [may not be true on unusual hardware](https://buildd.debian.org/status/fetch.php?pkg=pandas&arch=riscv64&ver=1.4.3%2Bdfsg-1&stamp=1659694881&raw=0).  (This was how I originally noticed the problem, as Debian tries to build packages on a wide range of hardware.)

This patch fixes both issues.  (In the Debian logs linked here, search for 'near-limits test', the first run is unpatched 1.3.5, the second run is patched 1.5.x.)

Its effect on speed is unclear: it is [sometimes faster](https://buildd.debian.org/status/fetch.php?pkg=pandas&arch=amd64&ver=1.5.1%2Bdfsg-1&stamp=1667870121&raw=0) and [sometimes slower](https://buildd.debian.org/status/fetch.php?pkg=pandas&arch=amd64&ver=1.5.2%2Bdfsg-1&stamp=1669859491&raw=0), possibly because other load on the build machines varies.

- [this bug does not appear to be previously reported ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [X] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [N/A ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

rebecca-palmer committed:
add test for float to_datetime near overflow bounds

rebecca-palmer committed:
fix float to_datetime near overflow bounds

rebecca-palmer committed:
fix typo and formatting

rebecca-palmer committed:
fix formatting

rebecca-palmer committed:
fix test to not fail on rounding differences

rebecca-palmer committed:
Merge branch 'main' into datetimenearbounds

rebecca-palmer committed:
don't use approximate comparison on datetimes, it doesn't work

rebecca-palmer committed:
also can't convert datetime to float

rebecca-palmer committed:
match dtypes

MarcoGorelli commented on pandas/_libs/tslib.pyx:
```diff
@@ -294,14 +294,22 @@ def array_with_unit_to_datetime(
             iresult[mask] = 0
+            # for bounds checking, which can't use (integer) iresult * mult
+            # because it needs arithmetic overflow to not wrap around
             fvalues = iresult.astype("f8") * mult
```
First, thanks for doing this

Second, I haven't looked closely at this yet (but will do!), just wanted to ask - would this overlap with, or supersede, https://github.com/pandas-dev/pandas/pull/48705 ?

rebecca-palmer replied:
and fvalues is now float-input-only, so this possible waste of time no longer exists

MarcoGorelli commented on pandas/_libs/tslib.pyx:
```diff
@@ -313,11 +321,9 @@ def array_with_unit_to_datetime(
                 result = (iresult * mult).astype("M8[ns]")
 
             elif values.dtype.kind == "f":
-                fresult = (values * mult).astype("f8")
-                fresult[mask] = 0
```
Just for my understanding, is this just a refactor (reuse `fvalues`, as it was already calculated above)?

MarcoGorelli reviewed (commented):
If I've understood correctly:

Before:
- `values` would get cast to `i8` (thus losing some precision, e.g. `292.32702463` -> `292`)
- the `i8` cast would then get cast back to `f8`, but once this has been done, the result could be smaller than what you started with, because of the loss of precision
- the bounds check might (erroneously) not pass

Whereas now:
- `values` are multiplied directly with `mult`
- as there's no loss of precision, the out-of-bounds check no longer erroneously passes

That's one change. The other change is to use `values != values` in the mask, as was done in #48705. Perhaps let's merge that one first, then rebase this one?

I've just left a question, as one change looks good but orthogonal (it's fine, just checking)

MarcoGorelli reviewed (changes requested):
https://github.com/pandas-dev/pandas/pull/48705 is in - could you rebase please and we can get this in too?

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

MarcoGorelli commented:
I'll add a commit here when I get a chance (hopefully early next week), I think we should ship this, and @rebecca-palmer did some seriously good work here

rebecca-palmer committed:
Merge branch 'main' into datetimenearbounds

rebecca-palmer committed:
TST: don't try to use non-integer years (see #50301)

jbrockmendel commented:
> Unrelated to my changes, but it looks like https://github.com/pandas-dev/pandas/pull/50263 may have accidentally undone the effect of https://github.com/pandas-dev/pandas/pull/50301 for some input types - @jbrockmendel ?

That's entirely plausible.  Would be happy to see a PR fixing that.

rebecca-palmer committed:
Merge branch 'main' into datetimenearbounds

rebecca-palmer commented:
That CI fail looks like the same one main then had - retrying.

MarcoGorelli reviewed (commented):
Looks good to me - only minor comments would be to add a whatsnew note in the 2.0.0 bug fixes section, and ideally to have a link to the github issue (or to the pull request if there's no issue) in the test (check some other tests for examples)

rebecca-palmer committed:
TST: don't cross an integer

(tsmax_in_days happens to be close to an integer,
and this is a test of rounding)

rebecca-palmer committed:
PERF: remove unnecessary copy

rebecca-palmer committed:
add whatsnew

rebecca-palmer commented on pandas/_libs/tslib.pyx:
```diff
@@ -294,14 +294,22 @@ def array_with_unit_to_datetime(
 
+        if values.dtype.kind in ["f",]:
+            mask = (values != values) | (values == NPY_NAT) # first is NaNs
+            fvalues = (values * mult).astype("f8")
```
Now uses copy=False

rebecca-palmer committed:
Merge branch 'main' into datetimenearbounds

rebecca-palmer commented:
Fixed, but also removed the unnecessary copy, so wait for the CI to check that it really is unnecessary.

MarcoGorelli reviewed (approved):
Looks good to me pending green, thanks @rebecca-palmer !

rebecca-palmer commented:
Does the Windows CI timing out suggest there's a hang bug in this, or does that just happen sometimes?

(Numpy Dev also fails on main, for reasons that are nothing to do with pandas: numpy/numpy#23033)
